Adding Analyze command to guacone

arorasoham9 opened the pull request:
# Description of the PR

Analyze command for guacone
`Comparing SBOMs with only patch version differences`
`guacone analyze diff --uri --sboms=https://anchore.com/syft/image/k8s.gcr.io/kube-apiserver-v1.24.1-583a02ce-8f7e-4794-91af-35f27ffeb73d,https://anchore.com/syft/image/k8s.gcr.io/kube-apiserver-v1.24.2-ee7e0a81-87de-4761-9689-4f7162d81e44`
<img width="1446" alt="Screenshot 2024-11-19 at 10 01 36 AM" src="https://github.com/user-attachments/assets/878b3cc8-a79b-459a-88e7-e8509f076779">



The diff is a 4 step process: 

- Create graph for SBOMs
- Find all paths in both graphs
- Find the paths that are present in one graph and not in the other, do this for both graphs.
- Pair paths from one graph to the other that are closest, then find the difference between these pairs.

Each step has tests to ensure that it is stable. Meaning, if the steps are repeated for the same Input SBOM, the results are the same. The JSON you see are the test SBOMS I use to carry out the tests for each step of the diff. These JSON files are marshalled HasSBOM nodes for SBOMs I ingested into GUAC, pulled using the Node function.

# PR Checklist

- [ ] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [ ] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If OpenAPI spec is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [ ] All dependent PRs have already been merged

arorasoham9 commented:
@pxp928 ready for review.

pxp928 commented:
Cool, @arorasoham9 can you add some screenshots to what the output will look like for the CLI?

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,666 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalyze
```
one major task would be to add unit tests to this package. Some things like the graph, getting data from graphQL can be abstracted away but do you have some tests you can add here to test the functionality?

arorasoham9 replied:
I don't have any at the moment. I can write a few.  Could they come in a future PR?

arorasoham9 replied:
Added unit tests as requested.

arorasoham9 commented:
> Cool, @arorasoham9 can you add some screenshots to what the output will look like for the CLI?

Added some screenshots.

jeffmendoza commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,121 @@
+  "github.com/guacsec/guac/pkg/logging"
+  "github.com/spf13/cobra"
+  "github.com/spf13/viper"
+  analysis "github.com/guacsec/guac/pkg/analyzer"
```
Any reason for the import rename here?

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+    node *Node
+  )
+  if node, err = g.Vertex(ID); err !=  nil {
+    fmt.Println("Error setting node attribute", err)
```
need to return this as an error.

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+  )
+  if node, err = g.Vertex(ID); err !=  nil {
+    fmt.Println("Error setting node attribute", err)
+    os.Exit(1)
```
don't use `os.Exit(1)` should return error

jeffmendoza replied:
Shouldn't have any os.Exit in a package under `pkg/`, assume this is a library that can be called from anywhere. Propagate errors back to the caller.

jeffmendoza commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,121 @@
+    if diff && intersect && union || diff && intersect || diff && union || intersect && union {
+      fmt.Println("Must specify only one of --diff, --intersect, --union")
+      return
+    }
```
This should be a positional argument if required, not a flag/option.

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+
+func GetNodeAttribute(g graph.Graph[string, *Node],ID, key string) interface{} {
+  var (
+    err error
```
no need to instantiate this here, can just it on line 72?

jeffmendoza commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    //create graphs
+    graphs := analysis.HasSBOMToGraph(cmd, ctx, gqlclient)
```
Don't pass cmd to a package under `pkg/`. Those shouldn't have a dependency on cli commands or cobra. Explicitly pass the parameters this needs.

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+  node.Attributes[key] = value
+}
+
+func GetNodeAttribute(g graph.Graph[string, *Node],ID, key string) interface{} {
```
why is this returning an `interface{}`

jeffmendoza commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,121 @@
+
+func init() {
+
+  rootCmd.PersistentFlags().Bool("intersect", false, "compute intesection of given sboms")
```
Don't add flags to rootCmd, this will show up on all `guacone` cmds if so. Use this `analyzeCmd` you are creating.
Also, cli options are normally under `pkg/cli/store.go` but this command has so many it might make sense to just leave them here.

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+    foundHasSBOMPkg, err = model.HasSBOMs(ctx, gqlclient, model.HasSBOMSpec{Subject: &model.PackageOrArtifactSpec{Package: &model.PkgSpec{Id: &pkgResponse.Packages[0].Namespaces[0].Names[0].Versions[0].Id}},
+      })
+    if err != nil {
+      fmt.Printf("(purl)failed getting hasSBOM with error :%v", err)
```
dont use `fmt.Printf`, you should return an error via `fmt.Errorf`

jeffmendoza commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,121 @@
+  rootCmd.PersistentFlags().Bool("purl", false, "input is a pURL")
+  rootCmd.PersistentFlags().Bool("id", false, "input is an Id")
+  rootCmd.PersistentFlags().Bool("metadata", false, "Compare SBOM metadata")
+  rootCmd.PersistentFlags().Bool("inclSoft", false, "Compare Included Softwares")
```
change to `incl-soft` and similar for others. All cli options use hyphen-case.

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+func verifyAnalyzeFlags(slsas, sboms []string,  errSlsa, errSbom error, uri, purl, id bool) {
+
+  if (errSlsa != nil && errSbom != nil) || (len(slsas) ==0  && len(sboms) == 0 ){
+    fmt.Println("Must specify slsa or sboms ")
```
return a proper error and dont use `os.Exit`

jeffmendoza commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+//
+// Copyright 2022 The GUAC Authors.
```
use current year for new files (unless you started writing this in 2022)

jeffmendoza commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalyze
```
package should be same name as directory.

jeffmendoza commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,121 @@
+  rootCmd.PersistentFlags().Int("maxprint", PRINT_MAX, "max number of similar sboms to print")
+  rootCmd.AddCommand(analyzeCmd)
+
+}
```
Add a newline to the end of the file

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+      namespaceTwo, okTwo := GetNodeAttribute(analysisGraph,diffList.MissingAddedRemovedLinks[i][1], "Namespace[0]").(string)
+
+      if !okOne || !okTwo {
+        fmt.Println("Error getting node namespace attribute")
```
clean up and pass out proper error

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+  table.SetAlignment(tablewriter.ALIGN_LEFT)
+  table.Render()
+  if (!all && max > maxprint){
+    fmt.Println("Run with --all to see full list")
```
should user logger here: https://github.com/guacsec/guac/blob/main/pkg/handler/processor/process/process.go#L117

It should be able to be pulled from the context:

```
logger := logging.FromContext(ctx)
```

jeffmendoza commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+  return foundHasSBOMPkg, nil
+}
+
+func verifyAnalyzeFlags(slsas, sboms []string,  errSlsa, errSbom error, uri, purl, id bool) {
```
All this should be under the file in `cmd/`, shouldn't have any flag validation here in `pkg/`.

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+}
+
+func HasSBOMToGraph(cmd *cobra.Command, ctx context.Context, gqlclient graphql.Client) ( []graph.Graph[string, *Node]){
+  slsas, errSlsa := cmd.Flags().GetStringSlice("slsa")
```
pass in options or something similar and not the actual `cmd`. See: https://github.com/guacsec/guac/blob/main/cmd/guacone/cmd/s3.go#L80-L93

pxp928 commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+  if uri {
+    hasSBOMResponseOne, err = FindHasSBOMBy(model.HasSBOMSpec{} ,sboms[0],"", "", ctx, gqlclient)
+    if err != nil {
+      fmt.Println("(uri)failed to lookup sbom:", sboms[0], err)
```
fix error

jeffmendoza commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+  //print to stdout
+  printHighlightedAnalysis(dot, diffList, all, maxprint, action, analysisGraph )
+}
+func max(nums []int) int {
```
add an empty line between functions.

pxp928 commented on pkg/analyzer/analyze_test.go:
```diff
@@ -0,0 +1,66 @@
+package guacanalyze_test
```
missing header

jeffmendoza commented on pkg/analyzer/analyze.go:
```diff
@@ -0,0 +1,665 @@
+  //Create dot file
+  createGraphDotFile(dot, analysisGraph)
+  //print to stdout
+  printHighlightedAnalysis(dot, diffList, all, maxprint, action, analysisGraph )
```
Anything under `pkg/` should not print anything to stdout. Any logic should result in a datastructure that returns back to `cmd/` and is then printed. This helps for testability. Looks like much of this may need to be moved back to `cmd/`

jeffmendoza reviewed (commented):
Mostly code organization and naming comments.

pxp928 commented:
@arorasoham9 did the `node` functionality get fixed?

arorasoham9 commented:
@pxp928  Ready for review. Please let me know what changes you require to the output presentation.

nathannaveen reviewed (commented):
Hey, @arorasoham9, I have a suggestion. Could you move the test data to https://github.com/guacsec/guac-test? 

There are over 1,000,000 lines of code in this PR, making it virtually impossible to review, and without adequate reviews, security bugs could be introduced into the code. 

If the test files were to be moved into guac-test, then the majority of this code wouldn’t be in this PR so that we can properly review it. 

If the test files were moved to guac-test, this code would not be built and intern would not affect the end user.

I am not entirely sure whether the JSON files are part of executable code. Even if they aren't right now I am not sure whether we will in the future. 

While JSON files are typically safe, there's a potential risk that they could contain encoded scripts, which might inadvertently introduce security vulnerabilities.

arorasoham9 commented:
@nathannaveen Done! I have moved the HasSBOM JSON test files for gaucanalyze to guac-test. I have also added some explanation for having these JSON files to the description of this PR.

I will require [this](https://github.com/guacsec/guac-test/pull/1)  PR merged before I can send a final commit to reflect the code changes to accept the JSON test from guac-test.

pxp928 commented:
@arorasoham9 any luck in improving the output? It is very hard to understand.

arorasoham9 commented:
@pxp928 I am still thinking of ways to show only the relevant information. This is a better explanation for why you see so much even with just a small difference between two sboms. The diff works at the granularity of a path, not a node, so when it finds even one node different it classifies all paths with that participating node as different (and prints them). I believe the original aim to find missing/different paths bw two sboms was to also offer a graph "structural" difference. I am trying to think how we can still offer the same differences in an easier to understand manner.

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+		uri, _ := cmd.Flags().GetBool("uri")
+		purl, _ := cmd.Flags().GetBool("purl")
+
+		metadata, _ := cmd.Flags().GetBool("metadata")
```
these flags should be initialized in https://github.com/guacsec/guac/blob/main/pkg/cli/store.go, and preferably prefixed in a way that shows its for guacanalyze

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+
+		if test == "" {
+			if err = verifyAnalyzeFlags(slsas, sboms, errSlsa, errSbom, uri, purl, id); err != nil {
+				fmt.Fprintf(os.Stderr, "Error: %s", err)
```
nit: golang errors style guide is lowercase first characters.

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+
+			//create graphs
+			graphs, err = hasSBOMToGraph(ctx, gqlclient, sboms, AnalyzeOpts{
+				Metadata: metadata, InclSoft: inclSoft, InclDeps: inclDeps, InclOccur: inclOccur, Namespaces: namespaces, URI: uri, PURL: purl, ID: id})
```
make it into different lines for readability.

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+				os.Exit(1)
+			}
+
+			//create graphs
```
nit: the comment doesn't add much context to the code, either remove or expand.

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+			}
+		}
+
+		if args[0] == "diff" {
```
would it make sense for these to be subcommands? otherwise this should be a flag maybe?

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+
+			if len(row) != 0 {
+				row = append(row, "--->")
+				table.SetColMinWidth(i+1, 90)
```
having constants for the formatting will make it easier to change in the future.

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+	}, nil
+}
+
+func verifyAnalyzeFlags(slsas, sboms []string, errSlsa, errSbom error, uri, purl, id bool) error {
```
nit: validate instead of verify for consistency https://github.com/guacsec/guac/blob/main/cmd/guaccollect/cmd/scorecard.go#L124

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+		gqlclient := graphql.NewClient(viper.GetString("gql-addr"), &httpClient)
+
+		slsas, errSlsa := cmd.Flags().GetStringSlice("slsa")
+		sboms, errSbom := cmd.Flags().GetStringSlice("sboms")
```
what is the err case you're expecting for this? would it be easier to have a default value and not worry about the errors? or is there a distinct behavior you're looking for?

lumjjb commented on pkg/analyzer/analyzer_test.go:
```diff
@@ -0,0 +1,154 @@
+
+var diffTestFile = "https://raw.githubusercontent.com/guacsec/guac-test/main/hasSbom-pairs/hasSBOM-syft-spdx-k8s.gcr.io-kube-apiserver.v1.24.1.json"
+
+func readTestFileFromHub(fileUrl string) ([]byte, error){
```
hmm how about using https://pkg.go.dev/embed here? with test files - can we get a smaller example to test with?

lumjjb commented on pkg/analyzer/analyzer.go:
```diff
@@ -0,0 +1,1009 @@
+			return g, fmt.Errorf("error setting metadata attribute(s)")
+		}
+	}
+	//TODO: inclSoft and inclOccur
```
can you create issues for each TODO and link them here?

lumjjb commented on pkg/analyzer/analyzer.go:
```diff
@@ -0,0 +1,1009 @@
+	return paths, nil
+}
+
+func HighlightAnalysis(gOne, gTwo graph.Graph[string, *Node], action int) ([][]*Node, [][]*Node, error) {
```
use enums for action can help readability

lumjjb commented on pkg/analyzer/analyzer.go:
```diff
@@ -0,0 +1,1009 @@
+func HighlightAnalysis(gOne, gTwo graph.Graph[string, *Node], action int) ([][]*Node, [][]*Node, error) {
+	pathsOne, errOne := FindPathsFromHasSBOMNode(gOne)
+	pathsTwo, errTwo := FindPathsFromHasSBOMNode(gTwo)
+	var analysisOne, analysisTwo [][]*Node
```
declare closer to use

lumjjb commented on pkg/analyzer/analyzer.go:
```diff
@@ -0,0 +1,1009 @@
+	return n.ID
+}
+
+func SetNodeAttribute(g graph.Graph[string, *Node], ID, key string, value interface{}) bool {
```
can you provide additional context and documentation around the functions? it is not clear what should be in value, unless there's a good reason, i would highly recommend not having `interface{}` as the type of value. If anything it can be a hidden type that can be minted by the analyzer but must provide sufficient constructors to make sense of it.

lumjjb commented on pkg/analyzer/analyzer.go:
```diff
@@ -0,0 +1,1009 @@
+
+	switch action {
+	//0 is diff
+	case 0:
```
use an enum :)

lumjjb commented on pkg/analyzer/analyzer.go:
```diff
@@ -0,0 +1,1009 @@
+	return hex.EncodeToString(hash[:])
+}
+
+func AddGraphNode(g graph.Graph[string, *Node], _ID, color string) {
```
what is color used for? Is it the identity of an edge of just for analysis purpose? can it be passed it as an option on the node for extensibility? 

`_ID` should be id to follow golang style?

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+	},
+}
+
+func printDiffedPathTable(diffs []analyzer.DiffedPath) error {
```
would be helpful to move all printing stuff to either a different package or a different file

lumjjb commented on cmd/guacone/cmd/analyze.go:
```diff
@@ -0,0 +1,476 @@
+	return nil
+}
+
+func readTwoSBOM(filename string) ([]graph.Graph[string, *analyzer.Node], error) {
```
why read two sboms from a file? i don't think that's a common pattern?

lumjjb reviewed (changes requested):
Thanks @arorasoham9 for working on this big feature!! definitely going to be very useful, would you be able to help break this up into a few separate PRs, that would be very helpful in getting this merged in faster! 

 maybe start with the analyzer package, and then add a table formatter pkg, and then the CLI options and the CLI itself?

lumjjb commented:
Hey @arorasoham9 , would you be able to chat at bit more amount helping to turn the corner to complete the PR.

arorasoham9 commented:
Reopen attempt
